libraries/chain/block_header_state.cpp

```diff
@@ -406,6 +401,8 @@ namespace eosio { namespace chain {
       additional_signatures = std::move(sigs);
 
       verify_signee();
+
+      h->producer_signature = header.producer_signature;
```

Comment:
Why pass in signed_block_header for the same object? This should have already been set on line 398.

Reply:
The `h` here is coming from producer_plugin.cpp:2637 which is actually a member of `block_state`, so they are different. However, it seems unnecessary to convert from `signed_block_ptr` to `std::share_ptr<signed_block_header>` for the declaration of `h`.

Reply:
`block_header_state.hpp` is unaware of the `signed_block` type.  It only knows about `signed_block_header`, for which a typedef/using shared pointer doesn't exist.

Reply:
Then have it take a `signed_block_header& h` instead.